chia/wallet/wallet_node.py

```diff
@@ -626,21 +630,28 @@ async def receive_and_validate(inner_states: List[CoinState], inner_idx_start: i
                     if len(valid_states) > 0:
                         self.log.info(
                             f"new coin state received ({inner_idx_start}-"
-                            f"{inner_idx_start + len(inner_states) - 1}/ {len(items) + 1})"
+                            f"{inner_idx_start + len(inner_states) - 1}/ {len(items)})"
                         )
                         assert self.new_state_lock is not None
                         async with self.new_state_lock:
+                            if self.wallet_state_manager is None:
+                                return
                             await self.wallet_state_manager.new_coin_state(valid_states, peer, fork_height)
-                        if update_finished_height:
-                            await self.wallet_state_manager.blockchain.set_finished_sync_up_to(
-                                last_change_height_cs(inner_states[-1])
-                            )
+
+                            if update_finished_height:
+                                if len(concurrent_tasks_cs_heights) == 1:
+                                    # We have processed all past tasks, so we can increase the height safely
+                                    synced_up_to = last_change_height_cs(valid_states[-1]) - 1
+                                else:
+                                    # We know we have processed everything before this min height
+                                    synced_up_to = min(concurrent_tasks_cs_heights)
```

Comment:
Let's say you start up tasks for 100-199, 200-299, and 300-399.  The first one fails out.  The third one gets here.  The minimum is going to be 200.  We have not process 100-199 but this code will set the finished sync up to to 200.  Right?

Reply:
If a task fails, then we will not add that state. This is an issue with trusted sync.
WIth untrusted, it's fine, since we always sync from scratch from new peers so we will catch it in the next sync. 

Not sure how to easily solve this problem, maybe another queue for failed CoinStates?

Reply:
How about a sorted list. when anything leaves it puts it's group in the list and then checked the already synced to height vs. the first element in the sorted list.  If that group is the next one after the existing height we can move forward.  Loop until either the end of the list or the first non-contiguous item (can use the same check as at first).  Probably better not to remove leading elements so figure out how many are to be removed then `the_list[:] = the_list[x:]` to mutate the existing list in one go.

https://docs.python.org/3/library/bisect.html#bisect.insort_right

Reply:
This is to support handling states that failed to add? I'm not sure about this, there are many reasons why a state might not add. Perhaps it's an invalid state, or there is a bug (or duplicate). I would just skip it or add it to some other queue to reprocess later. I think we should not handle it in this PR.